### PR TITLE
cmd/install: improve logic for following aliases

### PR DESCRIPTION
**Pull request by vladshablinsky**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

---

Changes for `cmd/install`:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/install.rb#L138-L139
`f.installed?` means that last prefix is installed for formula spec. However it seems to me that here it's treated as a check for any version being installed.

Before the change `f.installed?` did everything properly, it just refused to install formula if its latest prefix was installed. Not sure it was right to do so, but if `f.installed?` check had failed and formula still had been installed then `formula_installer` would have handled this.
Here is the diff with that change:
https://github.com/Homebrew/brew/pull/971/files#diff-b82c90101b7f1236459d3ba357ccc6e0

Also, 
`current ||= f.old_installed_formulae.first` doesn't always mean that `f.old_installed_formulae.first.installed?` is true.

These lines confuse me a little. As for now we can have formula installed and its revision changed, which means that `f.installed?` is false and let's suppose `f.old_installed_formulae.first` is not `nil`. 

The fix is to check if `f.installed?` is true or old formulae installed
with `f`'s current alias is `installed?` which implies that its latest prefix
is installed.

Maybe this is not the best fix for the issue, but I think we'll figure out what is best and then I'll update the PR.

CC @MikeMcQuaid @alyssais

### CODE REVIEW

**Comment by alyssais** [comment]
Yeah, I think what must have happened is that I assumed that `Formula#installed?` whether _any_ version was installed, rather than just the latest one — I must have figured we wouldn't want `brew install x` with any x installed to work, preferring `brew upgrade x` instead.

If that is the behaviour we want, the correct fix would be to replace the `f.installed?` check with one that checked whether any version was installed, and leave `old_installed_formulae` check as is.

However, it seems that the way this worked before was that `brew install x` when an old version of `x` was installed would just install the new version. If we want to keep it that way, we should just check `f.installed?` and skip any `old_installed_formulae` check, because if an old formula is installed, by definition it's not the current version.

**Comment by vladshablinsky** [comment]
> I must have figured we wouldn't want brew install x with any x installed to work, preferring brew upgrade x instead.

`Formula#installed?` can really confuse, so no worries. 

> However, it seems that the way this worked before was that brew install x when an old version of x was installed would just install the new version.

Yes, that's true if an old one is unlinked, otherwise it says that you need to unlink the old one to install new one. This behaviour is fine I guess, however I'm working on extending it so the user can install the latest version of any spec.

> If that is the behaviour we want, the correct fix would be to replace the f.installed? check with one that checked whether any version was installed, and leave old_installed_formulae check as is.

This check is performed in `formula_installer` for now and I'm not sure we need to change it right now, although it'll probably be changed when I finish with this: https://github.com/Homebrew/brew/issues/1162.

> leave old_installed_formulae check as is.

We need to check `old_installed_formulae` if new formula is not installed, right? AFAIU this change is done for better version support in future, could you elaborate why we need to perform this check at all?

Supposing we have the following mechanism for upgrading. does that mean that we get `boost@1.62 is already installed` or something when we're trying to install `boost@1.63`?

```
Upgrading a multiple version formula

In separate pull-requests:

Add the new keg_only version formula e.g. boost@1.63.rb
Modify the boost alias to point to boost@1.63.rb. Any formulae that need the old version of boost
should have their depends_on "boost" to be updated to depends_on "boost@1.62". boost@1.63.rb 
should no longer be keg_only and boost@1.62.rb should become keg_only.
When boost@1.63 has two major/minor versions newer than it (e.g. boost@1.65) then consider 
removing boost@1.63.rb and anything that depends on it.
```
(https://github.com/MikeMcQuaid/brew/blob/c3771336cff5630ce09248a57728e67393008087/docs/Migrate-A-Formula-To-Multiple-Versions.md)

**Comment by MikeMcQuaid** [comment]
> Yeah, I think what must have happened is that I assumed that Formula#installed? whether any version was installed, rather than just the latest one

I think many people make this assumption. It would be great to rename this method to `latest_installed?` or something and add `installed?` as a fallback to `compat`.

**Comment by MikeMcQuaid** [comment]
> Supposing we have the following mechanism for upgrading. does that mean that we get boost@1.62 is already installed or something when we're trying to install boost@1.63?

You should only get that if you do `brew install boost@1.62` not not for 1.63 through either `boost@1.63` or the `boost` alias.

File: Library/Homebrew/cmd/install.rb [hunk 1]
```diff
@@ -136,7 +136,12 @@ def install
         end
 
         current = f if f.installed?
-        current ||= f.old_installed_formulae.first
+        if current.nil? && f.installed_prefixes.empty?
+          old_formula_with_current_alias = f.old_installed_formulae.first
+          if old_formula_with_current_alias && old_formula_with_current_alias.installed?
+            current ||= old_formula_with_current_alias
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: Can make this `current =` as you've already verified it's `nil`.

File: Library/Homebrew/cmd/install.rb [hunk 2]
```diff
@@ -148,7 +153,7 @@ def install
           # Check if the formula we try to install is the same as installed
           # but not migrated one. If --force passed then install anyway.
           opoo "#{f.oldname} already installed, it's just not migrated"
-          puts "You can migrate formula with `brew migrate #{f}`"
+          puts "You should migrate formula with `brew migrate #{f}`"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: `You should migrate the formula with`

**Comment by alyssais** [comment]
> We need to check `old_installed_formulae` if new formula is not installed, right? AFAIU this change is done for better version support in future, could you elaborate why we need to perform this check at all?

`old_installed_formulae` is an extension of the check for old versions of the current formula. We should be checking one if and only if we're checking the other.

**Comment by alyssais** [comment]
This does look to me like the correct adjustment to account for alias changes while preserving the same behaviour. 👍

**Comment by MikeMcQuaid** [comment]
@vladshablinsky Any news here? Thanks!

**Comment by vladshablinsky** [comment]
@MikeMcQuaid let's close it for know and see if it's needed when #1343 merged.

**Pull request closed without merging**